source/includes/data-formats/extended-json.rb

```diff
@@ -0,0 +1,43 @@
+# start-ex-json-read
+require 'bson'
+
+ex_json = '''[
+    {"foo": [1, 2]},
+    {"bar": {"hello": "world"}},
+    {"code": {
+    "$scope": {},
+    "$code": "function x() { return 1; }"
+    }},
+    {"bin": {
+    "$type": "80",
+    "$binary": "AQIDBA=="
+    }}
+    ]'''
+
+doc = BSON::ExtJSON.parse(ex_json)
+
+puts doc.class
+puts doc
+# end-ex-json-read
+
+# start-ex-json-write
+require 'bson'
+
+hash_array = [
+    { "foo" => [1, 2] },
+    { "bin" => BSON::Binary.new("\x01\x02\x03\x04", :user) },
+    { "number" => BSON::Int64.new(42) }
+]
+
+json_array_canonical = hash_array.map(&:as_extended_json)
+json_array_relaxed = hash_array.map{ |hash| hash.as_extended_json(mode: :relaxed) }
+json_array_legacy = hash_array.map{ |hash| hash.as_extended_json(mode: :legacy) }
```

Comment:
Note that the bson library also defines `Array#as_extended_json`, which does pretty much what you have (mapping the array so each element has the method invoked on it as well).

(See https://github.com/mongodb/bson-ruby/blob/master/lib/bson/array.rb#L104-L108)

```suggestion
json_array_canonical = hash_array.as_extended_json
json_array_relaxed = hash_array.as_extended_json(mode: :relaxed)
json_array_legacy = hash_array.as_extended_json(mode: :legacy)
```